### ISSUE TITLE
Bidirectional attention mask in QwenOFT.

### ISSUE THREAD

**Issue by wyddmw** [starVLA/starVLA#97]
Hi, thanks for this awesome modular-based framework! According to the original OpenVLA-OFT implementation, the action tokens use a bidirectional attention mask instead of a causal attention mask for parallel decoding. I dig into the code in this repo, and I didn't find a specific attention modification, which means the default causal attention mask is adopted, if I'm not wrong. Since I'm a rookie to this area, I wonder whether a bidirectional attention mask is strictly required for parallel decoding, and what might be the difference in terms of the final performance. Due to the fact that flash-attention is widely adopted, a customized bidirectional attention mask might be hard to employ (lol).

**Comment by JinhuiYE** [comment]
Hi! Thank you for your interest in our framework and for the detailed observation.

You are correct in your understanding. In our implementation, we directly inherit the default settings of the action head from the base VLM, which means we adopt the **causal attention mask**, rather than the bidirectional attention mask used in the original OpenVLA-OFT for parallel decoding.

We didn’t aim to create a mirror replica of OpenVLA. Instead, our approach was to **copy and integrate the key idea** of how to insert the action module into a pre-trained VLM. As a result, there are some differences in specific design choices.

Regarding the impact on performance, based on our experimental experience so far, **using causal attention has not led to a substantial drop in performance**. We believe that aligning directly with the base VLM's attention mechanism is a more convenient and practical approach. It also makes the adaptation process from VLM to VLA more straightforward.

Thank you for bringing this up for discussion! If you have any other observations or ideas while using the framework, please feel free to share.

**Comment by wyddmw** [comment]
Hi, thanks for your reply! Your empirical conclusion is of great value. It's true that some representative vlas in autonomous driving, such as Simlingo, also adopt the default causal attention mask along with the parallel decoding strategy. The supervision on the action is also a straightforward L1/MSE loss instead of the shifted loss. This may lead to a gap between training and inference, but as you mentioned, the drop introduced by this gap may be acceptable. I will conduct my future research on this awesome repo, and looking forward to some potential collaborations with you (lol).
